### PR TITLE
[CDAP-20592] MessagingMetricsProcessorService crashes when adding delay topicDealy metrics

### PR DESCRIPTION
**Pull request by arjan-bal**
# [CDAP-20592](https://cdap.atlassian.net/browse/CDAP-20592) 

Pass a copy of metrics to persistMetricsAndTopicProcessMeta and use offer instead of add while appending delay metrics

[CDAP-20592]: https://cdap.atlassian.net/browse/CDAP-20592?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by tivv — COMMENTED** [review]
Please fix checkstyle

**Comment by arjan-bal** [comment]
> Please fix checkstyle

Done.

File: cdap-watchdog/src/main/java/io/cdap/cdap/metrics/process/MessagingMetricsProcessorService.java [hunk 1]
```diff
@@ -228,29 +229,38 @@ protected void run() {
...
 
-    // Persist metricsFromAllTopics and messageId's after all ProcessMetricsThread's complete.
-    // No need to make a copy of metricsFromAllTopics and topicProcessMetaMap because no thread is writing to them
-    persistMetricsAndTopicProcessMeta(metricsFromAllTopics, metadataHandler.getCache());
+    // Persist metricsFromAllTopics and messageId's after all
+    // ProcessMetricsThread's complete.
+    // Make a copy of metricsFromAllTopics because even though no thread is
+    // writing to it, the delay metrics are appended to the list. If we use
+    // the BlockingDeque, adding a metrics can cause queue size to be exceeded.
+    persistMetricsAndTopicProcessMeta(new LinkedList<>(metricsFromAllTopics),
```
Comment (tivv) [inline cdap-watchdog/src/main/java/io/cdap/cdap/metrics/process/MessagingMetricsProcessorService.java:244]: It's almost always better to use ArrayList or ArrayDeque (if you need queue notation)

**Review by tivv — COMMENTED** [review]
Please add a unit test

**Comment by arjan-bal** [comment]
Its not easy to repro this issue because after each iteration of fetching metrics from the messaging service, there's a call to [tryPersist](https://github.com/cdapio/cdap/blob/develop/cdap-watchdog/src/main/java/io/cdap/cdap/metrics/process/MessagingMetricsProcessorService.java#L374-L435z) which drains the queue if possible. I wasn't able to repro the issue and can't write a unit test to check this situation.

**Pull request closed without merging**